### PR TITLE
Application re-write

### PR DESCRIPTION
**Pull request by m-roberts**
## Functionality Description

Top level app has main loop running in thread, which waits when a user is using the display, and displays the image for the display context (known as a 'tile group') at the top of the stack. Switching to a new tile group is done by adding it to the app's stack, and returning to a previous tile group is done by popping.

There are 4 total tile groups:
* Bootsplash
* HUD (widget menu)
* Settings
* Screensaver

A "tile group" is just that - a list of bounding boxes that describe how a functionally distinct portion of an image is to be positioned within the displayed image. Tiles can be static, only showing text (for example), or can be dynamic, such as menus. Each tile can have multiple hotspots, which describe a particular object/region within the tile. Each hotspot is created in its own thread, which - if active (i.e. to be displayed) - calls its own render function with a given interval. Note: the interval can be dynamically reset, such as in the case of an animated GIF.

Button press events are handled by the app, which can respond to event announcements to make alterations to the top level tile group - this is how the state manager can invoke the screensaver, for example (i.e. by notifying the app to create a screensaver tile and push it onto the stack). The state manager performs additional button press handling, such as preventing user input during specific situations (e.g. running actions).

Events are handled by using module-level import observer pattern - this produces a singleton-like behaviour that allows for a single interface for responding to different state changes within the app, without higher-level classes requiring extended knowledge of other parts of the application. Optional data can be passed around in this way. Events are primarily used to handle button presses, start/stop actions/temporary tile groups (bootsplash/screensaver), notify the app that an active hotspot has updated its cached image (meaning the app should redraw the next image to the miniscreen), as well as navigating child/parent menus within a menu tile.

All scrolling is done using generators - the same mechanism is used for moving between menu pages as is used for scrolling text.

The general ownership of classes is like so:
```
App
    TileGroup
        Tile
            Hotspot
```

Let's use this image as an example (excuse the incorrect image):

This is the 'connection' page of the settings menu, which is managed by the settings TileGroup's menu tile. Each tile group uses its 'main' tile to handle button presses, etc. Menu tiles have a menu stack (similar to how the app has a tile group stack), which it uses for entering child menus. Child menus are created as required, similar to tile groups.

Menus have 4 navigation options, depending on context: up/down/child/parent. Some menu pages handle actions on SELECT as opposed to pushing a child menu to the menu tile's stack.

These have a status icon, and broadcast events (mentioned earlier) for changing state. If an action page has a 'get' method, then this used to determine if something is enabled/disabled:

If this cannot be determined, it is left in its default state of UNKNOWN. If a 'get method is not provided, then the icon has a 'play/submit' triangle:

## Other Changes
Other changes in this PR that are worth noting:

* Delete invalid `py.typed` notice file
* All images are now cropped to size and positioned precisely within the image
* Bootsplash is now a state
* Button presses are no longer handled in a stack
* Ignore build files in gitignore
* Exclude `legacy_widgets` dir (until reviewed!)
* Move `PT_MINISCREEN_SYSTEM` env var from systemd service to python app, so that it will always run correctly
* No absolute imports
* Remove `mode` from all objects (always use `"1"`)
* Logging output improved (`click_logging` module)
    * This includes a standardisation of how to invoke the log levels via CLI
* Top-level modularisation is improved - `main` creates and starts `App` instance
* Rename `pt-further-link.service` to `further-link.service` for bullseye
* Add `imgcat` support for rendering images to an `imgcat`-compatible terminal (useful for remote debugging)
* Rewrite `legacy_widgets` as menu pages/hotspots

-------

### General components
Some thought should be given to how the miniscreen onboarding can use generic components from this module. Event handling is the primary thing that needs to be considered, as the different pages have different events that the specific application is interested in - a better interface for this would be great!

### CODE REVIEW

**Comment by m-roberts** [comment]
> * 1 for Unguarded next in generator

https://lgtm.com/rules/11000086/

File: pt_miniscreen/app.py [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+            logger.debug(f"Current state: {self.state_manager.state}")
+
+            self.display(self.current_tile_group.image)
+            if environ.get("IMGCAT", "0") == "1":
+                print("\033c")
+                imgcat(self.current_tile_group.image)
+
+            logger.debug("Waiting until image to display has changed...")
+            start = time.time()
+            self.current_tile_group.wait_until_should_redraw()
```
Comment (m-roberts) [inline pt_miniscreen/app.py:185]: If we change tile group, are we still waiting on the previous one?

File: pt_miniscreen/app.py [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+
+    def _add_tile_group_to_stack_from_cls(self, tile_group_class) -> None:
+        self.add_tile_group(tile_group_class(size=self.miniscreen.size))
+
+    def setup_events(self) -> None:
+        subscribe(
+            AppEvent.START_BOOTSPLASH,
+            lambda: self._add_tile_group_to_stack_from_cls(PitopBootsplashTileGroup),
+        )
+        subscribe(AppEvent.STOP_BOOTSPLASH, lambda: self.pop_tile_group())
```
Comment (m-roberts) [inline pt_miniscreen/app.py:75]: if top of stack is instance `PitopBootsplashTileGroup`...

File: pt_miniscreen/hotspots/base.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+    def should_redraw(self):
+        if not self.draw_white and not self.draw_black:
+            return False
+
+        """Only requests a redraw after ``interval`` seconds have elapsed."""
+        return perf_counter() - self.last_updated > self.interval
+
+    @property
+    def image(self):
```
Comment (m-roberts) [inline pt_miniscreen/hotspots/base.py:47]: Does this get used? I think we use `paste_into` instead. I think `last_updated` might not be used either

File: pt_miniscreen/hotspots/base.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+        mask = None
+        if self.draw_white and not self.draw_black:
+            mask = hotspot_image
+
+        elif not self.draw_white and self.draw_black:
+            mask = ImageOps.invert(hotspot_image)
+
+        image.paste(hotspot_image, xy, mask)
+
+    def mask(self, hotspot_image):
```
Comment (m-roberts) [inline pt_miniscreen/hotspots/base.py:75]: Are we using this?

**Review by jcapona — APPROVED** [review]

**Pull request merged**